File: apps/ingestor/internal/event/bots.go
package event

import "regexp"

// botPattern matches common crawler/automation User-Agents. This is a pragmatic
// MVP filter (a superset of the obvious offenders); higher-fidelity bot
// detection is a later refinement — see the plan's polish phase.
var botPattern = regexp.MustCompile(`(?i)bot|crawl|spider|slurp|mediapartners|` +
	`bingpreview|facebookexternalhit|embedly|quora link preview|pinterest|` +
	`headless|phantomjs|puppeteer|playwright|selenium|` +
	`wget|curl|python-requests|python-urllib|go-http-client|java/|okhttp|axios|` +
	`apache-httpclient|libwww-perl|lighthouse|gtmetrix|pingdom|uptimerobot`)

// IsBot reports whether a request looks like a bot and should not be counted.
// An empty UA is treated as a bot (almost always automation).
func IsBot(userAgent string) bool {
	if userAgent == "" {
		return true
	}
	return botPattern.MatchString(userAgent)
}

File: apps/ingestor/src/lib/bots.ts
import { isbot } from 'isbot';

/** True if the request looks like a bot/crawler and should not be counted. */
export function isBot(userAgent: string | null): boolean {
	if (!userAgent) return true; // no UA → almost always automation
	return isbot(userAgent);
}

/** Extract the client IP from common proxy headers (transient; never stored). */
export function clientIp(headers: Headers): string {
	const fwd = headers.get('x-forwarded-for');
	if (fwd) return fwd.split(',')[0]!.trim();
	return (
		headers.get('cf-connecting-ip') ?? headers.get('x-real-ip') ?? '0.0.0.0'
	);
}
